src/include/mpir_tags.h

```diff
@@ -44,35 +44,32 @@
 #define MPIR_LAST_HCOLL_TAG           (MPIR_FIRST_HCOLL_TAG + 255)
 #define MPIR_FIRST_NBC_TAG            (MPIR_LAST_HCOLL_TAG + 1)
 
+#define MPIR_TAG_BITS_DEFAULT (31)
+
 /* These macros must be used carefully. These macros will not work with
  * negative tags. By definition, users are not to use negative tags and the
  * only negative tag in MPICH is MPI_ANY_TAG which is checked seperately, but
  * if there is a time where negative tags become more common, this setup won't
  * work anymore. */
 
+/* MPID_TAG_BITS should be declared by device later */
+#ifdef HAVE_TAG_ERROR_BITS
 /* This bitmask can be used to manually mask the tag space wherever it might
  * be necessary to do so (for instance in the receive queue */
-#ifdef HAVE_TAG_ERROR_BITS
-#define MPIR_TAG_ERROR_BIT (1 << 30)
-#else
-#define MPIR_TAG_ERROR_BIT
-#endif
-
+#define MPIR_TAG_ERROR_BIT (1 << (MPID_TAG_BITS-1))
 /* This bitmask is used to differentiate between a process failure
  * (MPIX_ERR_PROC_FAILED) and any other kind of failure (MPI_ERR_OTHER). */
-#ifdef HAVE_TAG_ERROR_BITS
-#define MPIR_TAG_PROC_FAILURE_BIT (1 << 29)
+#define MPIR_TAG_PROC_FAILURE_BIT (1 << (MPID_TAG_BITS-2))
+#define MPIR_TAG_ERROR_BITS (2)
```

Comment:
This change is related, but I'm leaning towards it being a separate commit prior to this commit. This change paves the way to making the tag space more dynamic.

Reply:
sorry I did not see this

Reply:
Just noting our discussion that this needs to be part of the commit that defines `MPID_TAG_BITS`, so my comment can be disregarded.